clockwork.go

```diff
@@ -85,27 +128,92 @@ type blocker struct {
 	ch    chan struct{}
 }
 
+func (s *sleeper) awaken(now time.Time) {
+	if atomic.CompareAndSwapUint32(&s.done, 0, 1) {
+		s.callback(s.arg, now)
+	}
+}
+
+func (s *sleeper) C() <-chan time.Time { return s.ch }
+
+func (s *sleeper) T() *time.Timer { return nil }
+
+func (s *sleeper) Reset(d time.Duration) bool {
+	active := s.Stop()
+	s.until = s.fc.Now().Add(d)
+	defer s.fc.addTimer(s)
+	defer atomic.StoreUint32(&s.done, 0)
+	return active
+}
+
+func (s *sleeper) Stop() bool {
+	stopped := atomic.CompareAndSwapUint32(&s.done, 0, 1)
+	if stopped {
+		// Expire the timer and notify blockers
+		s.until = s.fc.Now()
+		s.fc.Advance(0)
+	}
+	return stopped
+}
+
 // After mimics time.After; it waits for the given duration to elapse on the
 // fakeClock, then sends the current time on the returned channel.
 func (fc *fakeClock) After(d time.Duration) <-chan time.Time {
+	return fc.NewTimer(d).C()
+}
+
+// NewTimer creates a new Timer that will send the current time on its channel
+// after the given duration elapses on the fake clock.
+func (fc *fakeClock) NewTimer(d time.Duration) Timer {
+	done := make(chan time.Time, 1)
+	s := &sleeper{
+		fc:       fc,
+		until:    fc.time.Add(d),
+		callback: sendTime,
+		arg:      done,
+		ch:       done,
+	}
+	fc.addTimer(s)
+	return s
+}
+
+// AfterFunc waits for the duration to elapse on the fake clock and then calls f
+// in its own goroutine.
+// It returns a Timer that can be used to cancel the call using its Stop method.
+func (fc *fakeClock) AfterFunc(d time.Duration, f func()) Timer {
+	s := &sleeper{
+		fc:       fc,
+		until:    fc.time.Add(d),
+		callback: goFunc,
+		arg:      f,
+		// zero-valued ch, the same as it is in the `time` pkg
+	}
+	fc.addTimer(s)
+	return s
+}
+
+func (fc *fakeClock) addTimer(s *sleeper) {
 	fc.l.Lock()
 	defer fc.l.Unlock()
+
 	now := fc.time
-	done := make(chan time.Time, 1)
-	if d.Nanoseconds() == 0 {
+	if now.Sub(s.until) >= 0 {
 		// special case - trigger immediately
-		done <- now
+		s.awaken(now)
 	} else {
 		// otherwise, add to the set of sleepers
-		s := &sleeper{
-			until: now.Add(d),
-			done:  done,
-		}
 		fc.sleepers = append(fc.sleepers, s)
 		// and notify any blockers
 		fc.blockers = notifyBlockers(fc.blockers, len(fc.sleepers))
 	}
-	return done
+}
+
+func sendTime(c interface{}, now time.Time) {
```

Comment:
I usually try to put small functions like these inside the scope of whatever function needs it. For example:

```golang
func (fc *fakeClock) AfterFunc(d time.Duration, f func()) Timer {
        goFunc := func(fn interface{}, _ time.Time) {
                go fn.(func())() 
        }

	s := &sleeper{
  		fc:       fc,
		until:    fc.time.Add(d),
		callback: goFunc,
		arg:      f,
		// zero-valued ch, the same as it is in the `time` pkg
	}
	fc.addTimer(s)
	return s
}
```